### ISSUE TITLE
Proposal: customizing coroutine context switching

### ISSUE THREAD

**Issue by sshymko** [swoole/swoole-src#4197]
## The Problem

Swoole runtime is by design incompatible with FastCGI mechanisms. Global state specific to FastCGI is out of scope of Swoole. Such architecture is clean, but it greatly limits use cases of Swoole by applications/frameworks designed for PHP-FPM runtime. It is generally possible to retrofit such applications for compatibility with Swoole, but they remain conceptually incompatible with Swoole coroutines (because global context not managed by Swoole is shared between coroutines).

## Proposed Solution

Introduce events to track coroutine context switching with ability to preserve custom state.

**Events:**
- `WorkerSuspend` – fires _before_ the coroutine context switch
- `WorkerResume` – fires _after_ the coroutine context switch

### Request Context

Compatibility of super-global variables with Swoole coroutines.

```php
$server->on('request', function ($request, $response) {
    $_GET = (array)$request->get;
    $_POST = (array)$request->post;
    $_FILES = (array)$request->files;
    $_COOKIE = (array)$request->cookie;
    $_REQUEST = $_COOKIE + $_POST + $_GET;
    // ...
});

$server->on('workerSuspend', function () {
    // Return state associated with the current coroutine
    return [$_GET, $_POST, $_FILES, $_COOKIE, $_REQUEST];
});

$server->on('workerResume', function ($server, $workerId, $data) {
    // Restore state associated with the current coroutine
    [$_GET, $_POST, $_FILES, $_COOKIE, $_REQUEST] = $data;
});
```

### Sessions

Compatibility of the native PHP sessions with Swoole coroutines.

```php
$server->on('request', function ($request, $response) {
    session_start();
    try {
        // ...
    } finally {
        session_commit();
    }
});

$server->on('workerSuspend', function () {
    $data = [session_id(), $_SESSION];
    session_abort();
    return $data;
});

$server->on('workerResume', function ($server, $workerId, $data) {
    [$sessionId, $sessionData] = $data;
    session_id($sessionId);
    session_start();
    $_SESSION = $sessionData;
});
```

That will make libraries, such as [upscale/swoole-session](https://github.com/upscalesoftware/swoole-session), compatible with Swoole coroutines.

**Comment by twose** [comment]
Have you tried to make global vars be ArrayObject (or something that implements ArrayAccess)? Then you can get coroutine id and find coroutine context by it on ArrayObject::offsetGet().

I am not sure if it's ok to use session_* in coroutines, because your application may be blocked on it when it access the file system...

**Comment by sshymko** [comment]
@twose
Very clever idea to use `ArrayAccess` interface to scope the data by coroutine ID!
This should work at least in some cases, such as for the request super-globals variables.
How to retrieve the coroutine ID? And where to store the scoped values, in `$_GLOBALS`?

I'm afraid there going to be cases when scoping the data by coroutine ID is not going to be enough.
That includes every scenario when an environment cleanup is needed when switching between coroutines.
For instance, `session_abort()` call cleans up the process for subsequent call to `session_start()`.
The coroutine switching events are necessary to cover cases like this.

**Comment by twose** [comment]
> Very clever idea to use ArrayAccess interface to scope the data by coroutine ID!
This should work at least in some cases, such as for the request super-global variables.
How to retrieve the coroutine ID? And where to store the scoped values, in $_GLOBALS?

You can use `Coroutine::getContext()`, it returns ArrayObject and you can store anything in it.
You can implement context manager by yourself, I found an example here: https://github.com/hyperf/hyperf/blob/master/src/utils/src/Context.php

About session functions... I don’t like it, I would rather use the session I implemented... and I don't know much about it so I also don't know how to solve it...

**Comment by ValiDrv** [comment]
You should not really use those super globals even without Swoole.
But when you switch to Swoole, you need to wrap them.

Example:

```
class Context
{
    public static function set(string $key, mixed $value)
    {
        Co::getContext()[$key] = $value;
    }
    // Navigate the coroutine tree and find where we set `key`
    public static function get(string $key, mixed $default = null): mixed
    {
        $cid = Co::getcid();
        do {
            if (isset(Co::getContext($cid)[$key])) {
                return Co::getContext($cid)[$key];
            }
            $cid = Co::getPcid($cid);
        } while ($cid !== -1 && $cid !== false);

        # Not found
        return $default ?? throw new InvalidArgumentException("Could not find `{$key}` in context.", 404);
    }
}

...
$server->on('request', function ($request, $response) {
   Context::set('_GET', (array)$request->get);
   Context::set('_POST', (array)$request->post);
   Context::set('_FILES', (array)$request->files);
   Context::set('_COOKIE', (array)$request->cookie);
   ...
  # And when you use them
  echo Context::get('_GET')['foo'] ?? 'bar;
  # instead of
  echo $_GET['foo'] ?? 'bar;
});
```

If you do the context switching as you ask, it will slow down coroutines (and 99% of the time you don't need that data)

The only improvement I would do in Swoole would be that `Context::get` so it's closer to O1

**Comment by sshymko** [comment]
@ValiDrv
Thanks for providing a code example of what the Swoole coroutine context @twose was talking about!

Looks like the proposed approach will work for request super-globals. What would you recommend doing for `$_SESSION` that requires cleaning/restoring the session context (via `session_abort()` and `session_start()`) when coroutines switch?

**Comment by ValiDrv** [comment]
Normally you roll your own session, so you set it to some context variable on request, save it after the request.

So you don't call the PHP `session_` functions.

**Comment by fakharak** [comment]
@twose @ValiDrv 
I agree the proposal of @sshymko though not in the context of the "blocking" session, but for the transition of existing PHP-FPM based frameworks to Swoole's concurrency model.

### Problem Statement:
The problem with new concurrency mode in Swoole as set by `Co::set(['max_concurrency' => 1, ]);` is that while it makes the Worker Process behave more like PHP-FPM, the traditional PHP-FPM frameworks like "Laravel-S" and "Laravel Octane" strictly prohibit the use of Coroutine even inside Worker Processes (whereas Coroutine is the base of Asynchronous Programming Paradigm and without Coroutines the purpose of using Asynchrony dies).

These frameworks ab-use Swoole to provide hacks (way arounds) for the sake of over-cimplification. For example, Laravel Octane uses Task Workers for concurrency as a replacement of Coroutines which is wrong approach and against the core philosophy of Asynchrony.
 
 Laravel-S (on its Readme.md on github) clearly prohibits the use of Coroutines.
 
### Solution:
Instead of Swoole change its concurrency mode for traditional PHP-FPM based frameworks, a better solution could be an implicit Context Management (built internally in Swoole) such that traditional PHP-FPM based frameworks can specify those Singleton the data in which may be changed in different requests.

And, Swoole's built-in Context Manager implicitly inject the data of those Singleton and also the data of the Request, something like below:

$server->on('requestResume', function () {}); // Done implicitly by Swoole.

**Comment by ValiDrv** [comment]
I think that would defeat the purpose of those 3rd party libraries... since they MUST be made aware of the Swoole/Coroutines and if they were designed like that in the first place, you would not need this functionality.

And I think it will be very slow to re-set all the static/global variables on every coroutine change...

**Comment by sshymko** [comment]
+1 to @fakharak's proposal above!

The the ideal solution in the most general form is ability to switch arbitrary globals (mainly to replicate FastCGI/PHP-FPM behavior) when switching coroutines. Without such a mechanism, all attempts to retrofit "classical" frameworks/applications for compatibility with Swoole end up with disabled coroutines altogether which is ver unfortunate.

**Comment by fakharak** [comment]
> +1 to @fakharak's comment above!
> 
> The the ideal solution in the most general form is ability to switch arbitrary globals (mainly to replicate FastCGI/PHP-FPM behavior) when switching coroutines. Without such a mechanism, all attempts to retrofit "classical" frameworks/applications for compatibility with Swoole end up with disabled coroutines altogether which is very unfortunate.

Should "arbitrary globals" here mean "Only those globals (including singletons and services) which contain State (member data) and whose State is mutable with respect to each (user's) Request (which means whose state is NOT Constant / Same / Static for all requests but variable). (For the Worker Process, each Request runs as a Coroutine).

The Scenario:
When First Request (Coroutine) starts, we isolate globals variables through a "Coroutine Context Manager". When Second Request initiates, we do the same. When the First Request Resumes, it resumes the callback somewhere in the middle of code (from the point where Request / Coroutine halted for waiting for I/O. At this point, all Laravel's Routes / Singletons / Services (with State mutable per Request) would be expecting data from PHP Globals. Even if we somehow manage to initialize those Routes / Singletons / Services with Swoole\Http\Request then on resuming of First Coroutine all the values in Swoole\Http\Request would have been changed by the Second Request / Coroutine. 

Solution:
This means that all the Routes / Singletons / Services should also be initialized with (means, should depend on) the Request-specific data as isolated by Coroutine Context Manager.

The matter that remains to resolve is that "Who will do that initialization work ?"
A) Swoole only ? or,
B) Vendors of PHP-FPM Framework, only ? or,
C) Both Swoole and Framework Vendors (with the help of some kind of hooks / events system) ?

**Comment by fakharak** [comment]
What i understand is that the new changed Design Model for PHP-FPM based PHP-Frameworks (which currently run with "one-process-per-request" model) can be that the Route / Singletons / Services should now be "non-singleton" Objects created (locally) in callback for Swoole onRequest (local to the Request's coroutine-context) and should also be initialized in the same callback using the value stored in Coroutine Context of Request (which can be obtained using `ContextManager::get('key')`).

That proposal is based on assumption that when a Request (coroutine) resumes (after some I/O-task completes), then all of the objects, created (locally) inside that coroutine / context, become available again along with their actual State that they had before IO-Task started (that had handed over the execution to another request / coroutine).
